File: test/big_config/render_test.clj
(ns big-config.render-test
  (:require [big-config.keys :as k]
            [big-config.render :as render]
            [clojure.java.io :as io]
            [clojure.test :refer [deftest is use-fixtures]])
  (:import [java.nio.file Files]))

(defonce dirs (atom []))

(defn temp-dir []
  (let [dir (.toFile (Files/createTempDirectory "big-config-render-test-" (make-array java.nio.file.attribute.FileAttribute 0)))]
    (swap! dirs conj dir)
    (.getCanonicalPath dir)))

(defn delete-recursive [file]
  (let [f (io/file file)]
    (when (.exists f)
      (when (.isDirectory f)
        (doseq [child (.listFiles f)]
          (delete-recursive child)))
      (.delete f))))

(use-fixtures :each
  (fn [f]
    (try
      (f)
      (finally
        (doseq [d @dirs] (delete-recursive d))
        (reset! dirs [])))))

(deftest renders-templates-from-absolute-local-path
  (let [root (temp-dir)
        template (io/file root "template")
        target (io/file root "target")]
    (.mkdirs (io/file template "root"))
    (spit (io/file template "root" "hello.txt") "Hello {{ module }} {{ name }}")
    (let [res (render/render {k/render-module "infra"
                              k/render-templates [{:template (.getPath template)
                                                   :target-dir (.getPath target)
                                                   :overwrite true
                                                   :transform [["root"]]
                                                   :name "world"}]})]
      (is (= 0 (get res k/exit)))
      (is (= "Hello infra world" (slurp (io/file target "hello.txt")))))))

(deftest supports-function-sources-and-raw-transforms
  (let [root (temp-dir)]
    (render/copy-template-dir {:template-dir root
                               :target-dir (str (io/file root "target"))
                               :data {:module "infra"}
                               :src (fn [_key _data] "{{ module }}")
                               :files {:inventory "inventory.txt"}})
    (render/copy-template-dir {:template-dir root
                               :target-dir (str (io/file root "target"))
                               :data {:module "infra"}
                               :src (fn [_key _data] "{{ module }}")
                               :files {:raw "raw.txt"}
                               :opts [:raw]})
    (is (= "infra" (slurp (io/file root "target" "inventory.txt"))))
    (is (= "{{ module }}" (slurp (io/file root "target" "raw.txt"))))))

(deftest implements-selmer-whitespace-control-workaround
  (is (= "a{{ name }}b" (render/whitespace-control "a {{- name -}} b"))))

File: test/render.test.ts
import { mkdtempSync, readFileSync, rmSync, writeFileSync, mkdirSync } from "node:fs";
import { join } from "node:path";
import { tmpdir } from "node:os";
import { afterEach, describe, expect, it } from "vitest";
import { RENDER_MODULE, RENDER_TEMPLATES } from "../src/keys.js";
import { copyTemplateDir, render, whitespaceControl } from "../src/render.js";

const dirs: string[] = [];
function tempDir(): string {
  const d = mkdtempSync(join(tmpdir(), "big-config-render-test-"));
  dirs.push(d);
  return d;
}

afterEach(() => {
  for (const d of dirs.splice(0)) rmSync(d, { recursive: true, force: true });
});

describe("render", () => {
  it("renders templates from an absolute local dependency path", () => {
    const root = tempDir();
    const template = join(root, "template");
    const target = join(root, "target");
    mkdirSync(join(template, "root"), { recursive: true });
    writeFileSync(join(template, "root", "hello.txt"), "Hello {{ module }} {{ name }}");

    const res = render({
      [RENDER_MODULE]: "infra",
      [RENDER_TEMPLATES]: [{ template, "target-dir": target, overwrite: true, transform: [["root"]], name: "world" }]
    });
    expect(res["big-config/exit"]).toBe(0);
    expect(readFileSync(join(target, "hello.txt"), "utf8")).toBe("Hello infra world");
  });

  it("supports function sources and raw transforms", () => {
    const root = tempDir();
    copyTemplateDir({
      templateDir: root,
      targetDir: join(root, "target"),
      data: { module: "infra" },
      src: () => "{{ module }}",
      files: { inventory: "inventory.txt" }
    });
    copyTemplateDir({
      templateDir: root,
      targetDir: join(root, "target"),
      data: { module: "infra" },
      src: () => "{{ module }}",
      files: { raw: "raw.txt" },
      opts: ["raw"]
    });
    expect(readFileSync(join(root, "target", "inventory.txt"), "utf8")).toBe("infra");
    expect(readFileSync(join(root, "target", "raw.txt"), "utf8")).toBe("{{ module }}");
  });

  it("implements Selmer whitespace control workaround", () => {
    expect(whitespaceControl("a {{- name -}} b")).toBe("a{{ name }}b");
  });
});
